File: SEPTEMBER 23/13 September/135. Candy LC.cpp
class Solution {
public:
    int candy(vector<int>& ratings) {
        int n = ratings.size();
        vector<int> candies(n);
        vector<int> pre(n,0), suf(n,0);
        for(int i=1; i<n; i++)
        {
            if(ratings[i-1] < ratings[i])
                pre[i] = pre[i-1]+1;
        }
        for(int i=n-2; i>=0; i--)
        {
            if(ratings[i] > ratings[i+1])
                suf[i] = suf[i+1]+1;
        }     
        int totalCandies = 0;
        for(int i=0; i<n; i++)
            totalCandies+=max(pre[i],suf[i]);
        return totalCandies+n;
    }
};

File: JUNE 2025/02 June/135. Candy LC.java
class Solution {
    public int candy(int[] ratings) {
        int N = ratings.length;
        int[] candies = new int[N];
        Arrays.fill(candies, 1);
        // forward pass
        for(int i=1; i<N; i++)
        {
            if(ratings[i]>ratings[i-1])
                candies[i] = candies[i-1]+1;
        }
        // backward pass
        int totalCandies = candies[N-1];
        for(int i=N-2; i>=0; i--)
        {
            if(ratings[i]>ratings[i+1])
                candies[i] = Math.max(candies[i], candies[i+1]+1);
            totalCandies += candies[i];
        }

        return totalCandies;
    }
}
